### PR TITLE
Feature/backend functionality

### PR DESCRIPTION
**Pull request by SimonSchick**
I'm done completely replacing joi with jojen in the back-end, the current version passes all unit and integration tests.

I think this is worth a release, probably major.

### CODE REVIEW

File: lib/rules/number.js [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+    }
+
+    validateSync(params) {
+        const match = params.value.toString().match(precisionRegEx);
+        if (!match || !match[1]) {
+            return true;
+        }
+
+        const length = match[1].length - 1;
+        return length <= this._val || {
```
Comment (connor4312) [inline lib/rules/number.js]: This is off by one. The following test fails:

```
expect(Jo => Jo.number().precision(2)).to.failOn(42.424)
```

File: lib/rules/number.js [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+const precisionRegEx = /^-?\d+(\.\d+)?$/;
+
+class Precision extends SyncRule {
+
+    compile(params) {
+        this._val = params.args[0];
+    }
+
+    coerce(value) {
+        return value.toFixed(this._val);
```
Comment (connor4312) [inline lib/rules/number.js]: This converts the value into a string, which is not what should happen

File: lib/rules/object.js [hunk 3]
```diff
@@ -16,12 +14,23 @@ class ObjectValidator extends Rule {
 
 class Keys extends Rule {
     compile(params) {
+        const found = params.invokeFirst(this.constructor, rule => {
+            Object.assign(rule._obj, params.args[0]);
```
Comment (connor4312) [inline lib/rules/object.js]: Use assign from the util module. Object.assign isn't available everywhere yet.

File: lib/rules/string.js [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
+        return typeof params.value === 'string' &&
+            params.value.length > 0;
+    }
+
+    static ruleName() {
+        return 'string';
+    }
+}
+
+/* eslint-disable no-unreachable */
```
Comment (connor4312) [inline lib/rules/string.js]: This doesn't seem necessary

File: lib/rules/string.js [hunk 5]
```diff
@@ -0,0 +1,407 @@
...
+            let intVal = parseInt(str.substr(i, 1), 10);
+            if (i % 2 === 0) {
+                intVal *= 2;
+                if (intVal > 9) {
+                    intVal = 1 + (intVal % 10);
+                }
+            }
+            sum += intVal;
+        }
+        return (sum % 10) === 0;
```
Comment (connor4312) [inline lib/rules/string.js]: It'd be preferable to just import it from [isCreditCard](https://github.com/chriso/validator.js/blob/master/src/lib/isCreditCard.js), I don't see a reason to duplicate the code here. If you duplicate it you also have to include the license from Validator.js.

File: lib/rules/string.js [hunk 6]
```diff
@@ -0,0 +1,407 @@
...
+                if (matcher instanceof RegExp) {
+                    return matcher;
+                }
+                if (typeof matcher === 'string') {
+                    return new RegExp(`^${escapeRegExp(matcher)}$`);
+                }
+                return matcher;
+            });
+        }
+        this._allowRelative = options.allowRelative;
```
Comment (connor4312) [inline lib/rules/string.js]: Could be simplified as:

```
this._allowRelative = options.allowRelative;
this._schemes = (options.schemes || []).map( ... );
```

To reduce nesting

File: lib/rules/string.js [hunk 7]
```diff
@@ -0,0 +1,407 @@
...
+                    reason: 'relativ-not-allowed',
+                };
+            }
+            return false;
+        }
+        if (!isURL(params.value)) {
+            return false;
+        }
+        return true;
+    }
```
Comment (connor4312) [inline lib/rules/string.js]: The complexity in this function is fairly high, it can probably be broken up a bit

File: test/_setup.js [hunk 8]
```diff
@@ -2,36 +2,52 @@ const Joi = require('joi');
...
 
-chai.use(function (_chai, utils) {
-    Assertion.addMethod('failOn', function (value, details) {
+    return str;
+}
+
+const validFns = [];
+
+chai.use((_chai, utils) => {
+    Assertion.addMethod('failOn', function (value, details, ignoreCompability, options) {
```
Comment (connor4312) [inline test/_setup.js]: The arguments after `value` should probably all be moved into an options object.

File: test/validator.test.js [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+            }).not.to.throw();
+        });
+    });
+
+    describe('validateSync()', () => {
+        it('works correctly', () => {
+            expect(Jo.validateSync(1, Jo.number())).to.eql({
+                value: 1,
+                error: null,
+            });
```
Comment (connor4312) [inline test/validator.test.js]: Test with failing validators and with asynchronous validations to ensure it works

File: test/string.test.js [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+        expect(Jo => Jo.string().trim()).to.not.failOn(' AAAA ');
+    });
+
+    it('will validate iso dates', () => {
+        expect(Jo => Jo.string().isoDate()).to.not.failOn('1997');
+        expect(Jo => Jo.string().isoDate()).to.not.failOn('1997-07');
+        expect(Jo => Jo.string().isoDate()).to.not.failOn('1997-07-16');
+        expect(Jo => Jo.string().isoDate()).to.not.failOn('1997-07-16T19:20+01:00');
+        expect(Jo => Jo.string().isoDate()).to.not.failOn('1997-07-16T19:20:30+01:00');
+        expect(Jo => Jo.string().isoDate()).to.not.failOn('1997-07-16T19:20:30.45+01:00');
```
Comment (connor4312) [inline test/string.test.js]: Test with dates that _do_ fail

File: test/string.test.js [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+        expect(Jo => Jo.string().guid()).to.failOn(':D:DD:D:DD:');
+    });
+
+    it('will validate HEX', () => {
+        expect(Jo => Jo.string().hex()).to.not.failOn('af2a57c3fe14');
+        expect(Jo => Jo.string().hex()).to.failOn('zzzz');
+    });
+
+    it('will validate hostnames', () => {
+    });
```
Comment (connor4312) [inline test/string.test.js]: Missing tests?

File: test/string.test.js [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+
+    it('will validate emails', () => {
+        expect(Jo => Jo.string().email()).to.not.failOn('something@something.com');
+        expect(Jo => Jo.string().email()).to.failOn('something@@something.com');
+    });
+
+    it('will validate IPs', () => {
+        expect(Jo => Jo.string().ip()).to.not.failOn('0.0.0.0');
+        expect(Jo => Jo.string().ip()).to.failOn('0.0.0.0.0');
```
Comment (connor4312) [inline test/string.test.js]: IPv6? More cases?

File: test/any.test.js [hunk 13]
```diff
@@ -59,4 +61,14 @@ describe('any', () => {
         expect((Jo) => Jo.invalid(['a', 'b'])).to.failOn('b');
         expect((Jo) => Jo.invalid(['a', 'b'])).not.to.failOn('c');
     });
+
+    it('allows allowed values', () => {
+        expect((Jo) => Jo.allow('a')).to.not.failOn('c');
+        expect((Jo) => Jo.string().allow(1)).not.to.failOn(1);
```
Comment (connor4312) [inline test/any.test.js]: Check for multiple calls to allow as well

File: lib/rules/number.js (hunk 2 above)
Comment (SimonSchick) [inline lib/rules/number.js]: Jojen converts by default, if you add the `convert: false` option it will fail as you expect.

Comment (connor4312) [inline lib/rules/number.js]: Right, but it should coerce to a number of the give precision, not a string.

Comment (SimonSchick) [inline lib/rules/number.js]: Yes but that's a separate thing which I have already fixed locally.

Comment (SimonSchick) [inline lib/rules/number.js]: Oh wait, I commented on the wrong comment :-1:

**Comment by connor4312** [comment]
I think that's about it. @WatchBeam/devops have anything else to add?

File: package.json [hunk 14]
```diff
@@ -1,12 +1,13 @@
...
   "description": "Expressive validator for Node and the browser, with a small footprint and awesome performance",
-  "main": "lib/index.js",
+  "main": "dist/index",
   "scripts": {
+    "debug": "node debug mocha -r test/_setup.js --compilers js:babel-register && eslint lib",
     "test": "mocha -r test/_setup.js --compilers js:babel-register && eslint lib",
-    "sizeBrowser": "browserify lib/index.js -t babelify --presets es2015 | ccjs --compilation_level=ADVANCED_OPTIMIZATIONS | gzip | wc -c",
-    "prepublish": "babel lib --out-dir dist"
+    "bundle": "browserify lib/index.js -t babelify --presets es2015 | ccjs --compilation_level=ADVANCED_OPTIMIZATIONS > jojen.min.js",
+    "prepublish": "babel lib --out-dir dist && (type casdcjs &> /dev/null && npm run bundle || echo 'closure compiler not installed, not bundling for browser')"
```
Comment (connor4312) [inline package.json]: This breaks builds on Windows environments.

Comment (SimonSchick) [inline package.json]: Do you have any idea for alternatives? I don't use windows as dev environment.

Comment (connor4312) [inline package.json]: I would just put the closure compiler as an optional dependency and leave it at that.

File: .travis.yml [hunk 15]
```diff
@@ -4,3 +4,8 @@ node_js:
   - "5"
 git:
   depth: 10
+
+branches:
```
Comment (ttaylorr) [inline .travis.yml]: Hmm, does this disable test runs for pull-requests? I've always liked having that functionality.

Comment (connor4312) [inline .travis.yml]: Not for PRs, just branches!

**Pull request merged**